spmd/tensor/ops/math_ops.py

```diff
@@ -117,24 +117,37 @@ def einop_rule(
     )
 
 
-@register_impl("aten.sum.default")
-def dist_sum(self: DTensor) -> DTensor:
-    self_local = self.to_local()
-    self_placement = self.placements[0]
-    device_mesh = self.device_mesh
-
-    local_sum = self_local.sum()
-
-    if self_placement.is_shard() or self_placement.is_partial():
-        placements = [_Partial()]
-        # partial reduce
-        partial_sum = DTensor(local_sum, device_mesh, placements)
-        # all_reduce across device
-        replicate_placements = [Replicate()]
-        return partial_sum.redistribute(device_mesh, replicate_placements)
-    elif self_placement.is_replicate():
-        return DTensor(
-            local_sum, device_mesh=device_mesh, placements=self.placements
+def reduction_rule(op_schema: OpSchema) -> OutputSharding:
+    """
+    Propagate the sharding for reduction operations. Examples:
+        ij->i - sum on dim
+    """
+    alphabet = "abcdefghijklmnopqrstuvwxyz"
+    # reduction op usually begin with a single tensor
+    input_spec = cast(PlacementSpec, op_schema.args_schema[0])
+    input_chars = alphabet[: input_spec.ndim]
+
+    if len(op_schema.args_schema) > 1 and isinstance(
+        op_schema.args_schema[1], (int, list)
+    ):
+        # this is usually a dim-based reduction op schema pattern, it
+        # might not be true for every op for other special cases, we
+        # need to specialize them as needed.
+        # TODO: add support for things like `torch.unique` where it
+        # does not follow the reduction op convention.
+        dim_list = as_list(op_schema.args_schema[1])
+        out_dimchars = input_chars.translate(
```

Comment:
Interesting.. Python is magic!

Reply:
translate seems super fast compare with other string replacement after some search, so i use this haha, magic indeed.